Adds clip range for JointAction

fan-ziqi opened the pull request:
# Description

This PR adds support for action clip to all mdp/actions. Clip ranges can be specified as a dictionary of joint names and tuple for the lower and upper bounds of the clip in the ActionTermCfg.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update


## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

jtigue-bdai commented:
Currently limits are handled by the actuators. Primarily this is done on the effort.

fan-ziqi commented:
So how can I clip actions in isaaclab like in legged_gym?

jtigue-bdai commented:
Hmm I guess if you are using joint position commands the actuator limits won't really handle that. Actuator limits really only do effort. 

I will take another look at the current api and get back to you.

fan-ziqi commented:
Thanks for reply!
Yes, I use Δjoint_pos as actions

kellyguo11 commented:
The discussion in this PR may be relevant - https://github.com/isaac-sim/IsaacLab/pull/984. This seems like something that would make more sense for the wrappers.

fan-ziqi commented:
Which method is better? I don't know if this should be an env feature or an action feature

jtigue-bdai commented:
I guess this one only handles the joint_actions.

fan-ziqi commented:
So is this PR necessary? 
Should we adopt this or #984?

fan-ziqi commented:
For me, this allows more flexibility in configuring each joint individually.

So is this PR necessary?
Should we adopt this or #984?

kellyguo11 commented:
That's a good point, this could be useful for configuring the joints separately, but only applicable for joint actions. Perhaps there's value in adopting both approaches.

jtigue-bdai commented:
I could see both and also adding the clip configuration at the base ActionTerm/Cfg. It would just require that the actual processing be added to all provided actions.

jtigue-bdai commented:
Hey @fan-ziqi would you be willing to add the clipping cfg to the top level ActionTermCfg, the clip checking in the top level ActionTerm.__init__ , and then add the clipping execution to all the provided actions in the envs/mdp/actions directory?

fan-ziqi commented:
I'm working on that.

I have a question, is `scale: dict[str, tuple]` necessary? or `scale: tuple[float, float]` is enough

fan-ziqi commented:
I have added `clip` to `ActionTermCfg` and added clip action to all `mdp/actions`

But I have a question, is `scale: dict[str, tuple]` necessary? 

```python
@configclass
class ActionsCfg:
    """Action specifications for the MDP."""

    joint_pos = mdp.JointPositionActionCfg(
        asset_name="robot", joint_names=[".*"], scale=0.5, use_default_offset=True, clip={".*": (-100.0, 100.0)}
    )
```

or `scale: tuple[float, float]` is enough

```python
@configclass
class ActionsCfg:
    """Action specifications for the MDP."""

    joint_pos = mdp.JointPositionActionCfg(
        asset_name="robot", joint_names=[".*"], scale=0.5, use_default_offset=True, clip=(-100.0, 100.0)
    )
```

jtigue-bdai commented:
I think being able to do clip=dict[str,tuple] is necessary to be able to set clips on a per joint basis.

kellyguo11 commented:
Thanks a lot for the feature! would it be possible to add some unit tests for it as well?

kellyguo11 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/actions/binary_joint_actions.py:
```diff
@@ -115,6 +123,13 @@ def process_actions(self, actions: torch.Tensor):
         self._processed_actions = torch.where(binary_mask, self._close_command, self._open_command)
+        # clip actions
+        if self._clip is not None:
+            # resolve the dictionary config
```
is it possible to resolve this once at initialization and convert to a torch tensor to avoid more overhead with the loop?

fan-ziqi replied:
OK, but I encountered a problem with the code

```python
# parse clip
        if cfg.clip is not None:
            if isinstance(cfg.clip, dict):
                print(cfg.clip)
                self._clip = torch.tensor([[-float('inf'), float('inf')]], device=self.device).expand(self.num_envs, self.action_dim, 2)
                index_list, _, value_list = string_utils.resolve_matching_names_values(self.cfg.clip, self._joint_names)
                for idx, (min_val, max_val) in zip(index_list, value_list):
                    self._clip[:, idx] =
            else:
                raise ValueError(f"Unsupported clip type: {type(cfg.clip)}. Supported types are dict.")
```

I don't know how to put the value into `self._clip[:, idx]`, direct assignment will cause broadcasting problems

fan-ziqi replied:
solved.

fan-ziqi commented:
The preprocessing of clip has been moved to init @kellyguo11

kellyguo11 commented:
> The preprocessing of clip has been moved to init @kellyguo11

Thanks! Please also run the formatter and would be great if you can add some tests.

fan-ziqi commented:
> > The preprocessing of clip has been moved to init @kellyguo11
> 
> Thanks! Please also run the formatter and would be great if you can add some tests.

Code formatter has run.

I am not familiar with the unit testing of this project. If you need me to write one, can you give me a reference?

kellyguo11 commented:
> > > The preprocessing of clip has been moved to init @kellyguo11
> > 
> > 
> > Thanks! Please also run the formatter and would be great if you can add some tests.
> 
> Code formatter has run.
> 
> I am not familiar with the unit testing of this project. If you need me to write one, can you give me a reference?

Great, thanks! Unit tests are generally implemented here: https://github.com/isaac-sim/IsaacLab/tree/main/source/extensions/omni.isaac.lab/test. We can maybe copy one of the manager ones and apply it to the ActionManager with ActionTerms that have a clipping parameter.

fan-ziqi commented:
I noticed that there are no unit tests for `ActionManager`. I'm not familiar with it, and I have other work to handle at the moment, so I think I’ll have to leave it to you to add the tests.
